core/psh/nc.c

```diff
@@ -0,0 +1,337 @@
+/*
+ * Phoenix-RTOS
+ *
+ * nc - TCP and UDP connect or listen
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <sys/select.h>
+
+#include "psh.h"
+
+
+/*
+ * This could be achieved using getaddrinfo(),
+ * however it does not recognize scope id for now
+ * and requires dynamic memory, which we
+ * are trying to avoid.
+ */
+static int psh_nc_sockaddrFill(int *af_family, char *addrstr, char *portstr,
+                               struct sockaddr_storage *sockaddr,
+                               socklen_t *addrlen)
+{
+	struct sockaddr_in6 *saddr_in6;
+	struct sockaddr_in *saddr;
+	char *ifname, *endptr;
+	int port = 0, scopeid = 0;
+	uint32_t addr[4] = { };
+
+	if (*af_family == AF_UNSPEC) {
+		if (addrstr && strchr(addrstr, ':') != NULL)
+			*af_family = AF_INET6;
+		else
+			*af_family = AF_INET;
+	}
+
+	/* Try to resolve scope id */
+	if (*af_family == AF_INET6) {
+		ifname = strtok(addrstr, "%");
+		if ((ifname = strtok(NULL, "%")) != NULL) {
+			scopeid = strtol(ifname, &endptr, 0);
+			if (endptr == ifname) {
+				/* Assume it's a netif name */
+				if ((scopeid = if_nametoindex(ifname)) == 0) {
+					fprintf(stderr, "nc: No such interface: %s!\n", ifname);
+					return -1;
+				}
+			}
+			else if (scopeid <= 0) {
+				fprintf(stderr, "nc: Interface idx must be greater than 0!\n");
+				return -1;
+			}
+			/* Assume we got a netif idx */
```

Comment:
What if `*endptr != '\0'`

Reply:
Ok, we should handle this case to prevent strings, which don't consist of digits only.